### PR TITLE
Add, test and fix CryptoAEADAES256GCM functions

### PR DESCRIPTION
**Pull request by silkeh**
This pull requests extends the CryptoAEADAES256GCM functions and adds tests for the encryption and decryption as well.

For testing, I have used [google/gofuzz](https://github.com/google/gofuzz) for generating random data. This has as downside that the result of the detached encryption is not compared to a reference result (but the other functions are tested against it). An alternative is using the data that is used by libsodium, see [my other branch](/silkeh/libsodium-go/tree/crypto_aead_aes256gcm).

### CODE REVIEW

File: cryptoaead/crypto_aead_aes256gcm_test.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	Message    []byte
+	Ad         []byte
+	Key        [32]byte
+	Nonce      [12]byte
+}
+
+func TestCryptoAEADAES256GCM(t *testing.T) {
+	// Skip the test if unsupported on this platform
+	if !CryptoAEADAES256GCMIsAvailable() {
+		t.Skip()
```
Comment (redragonx) [inline cryptoaead/crypto_aead_aes256gcm_test.go]: This should log an warning for unsupported platforms. In case people assume it worked? Thoughts?

Comment (silkeh) [inline cryptoaead/crypto_aead_aes256gcm_test.go]: I have added some text to go with the skip.

**Comment by redragonx** [comment]
I like the idea of using random data, have you looked in different ports to see what people do the most?

**Comment by silkeh** [comment]
Most of the libraries I have looked at use a set of test vectors (including an expected ciphertext). The rest seem to use an input string and see if they can encrypt/decrypt it. This test essentially does the latter, but should also cover all edge cases that the test vectors cover. The downside is that the ciphertext from the *detached encryption* is only verified against the output of the *encryption* function. I think this should still catch most errors though.

**Comment by silkeh** [comment]
I have added the remaining functions from `crypto_aead_aes256gcm`.

**Comment by redragonx** [comment]
I will look at this pr in depth when I get a chance, thanks for your hard work.

**Comment by silkeh** [comment]
Take your time. I'll continue working on some more pull requests 😉.
Thanks for taking a look at the requests so quickly!
